scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -201,108 +247,221 @@ def _highly_variable_genes_single_batch(
+    disp_grouped = df.groupby("mean_bin", observed=True)["dispersions"]
     if flavor == "seurat":
-        df["mean_bin"] = pd.cut(df["means"], bins=n_bins)
-        disp_grouped = df.groupby("mean_bin", observed=True)["dispersions"]
-        disp_mean_bin = disp_grouped.mean()
-        disp_std_bin = disp_grouped.std(ddof=1)
-        # retrieve those genes that have nan std, these are the ones where
-        # only a single gene fell in the bin and implicitly set them to have
-        # a normalized disperion of 1
-        one_gene_per_bin = disp_std_bin.isnull()
-        gen_indices = np.where(one_gene_per_bin[df["mean_bin"].to_numpy()])[0].tolist()
-        if len(gen_indices) > 0:
-            logg.debug(
-                f"Gene indices {gen_indices} fell into a single bin: their "
-                "normalized dispersion was set to 1.\n    "
-                "Decreasing `n_bins` will likely avoid this effect."
-            )
-        # Circumvent pandas 0.23 bug. Both sides of the assignment have dtype==float32,
-        # but there’s still a dtype error without “.value”.
-        disp_std_bin[one_gene_per_bin.to_numpy()] = disp_mean_bin[
-            one_gene_per_bin.to_numpy()
-        ].to_numpy()
-        disp_mean_bin[one_gene_per_bin.to_numpy()] = 0
-        # actually do the normalization
-        df["dispersions_norm"] = (
-            df["dispersions"].to_numpy()  # use values here as index differs
-            - disp_mean_bin[df["mean_bin"].to_numpy()].to_numpy()
-        ) / disp_std_bin[df["mean_bin"].to_numpy()].to_numpy()
+        disp_stats = _stats_seurat(df["mean_bin"], disp_grouped)
     elif flavor == "cell_ranger":
-        from statsmodels import robust
+        disp_stats = _stats_cell_ranger(df["mean_bin"], disp_grouped)
+    else:
+        raise ValueError('`flavor` needs to be "seurat" or "cell_ranger"')
 
-        df["mean_bin"] = pd.cut(
-            df["means"],
-            np.r_[-np.inf, np.percentile(df["means"], np.arange(10, 105, 5)), np.inf],
-        )
-        disp_grouped = df.groupby("mean_bin", observed=True)["dispersions"]
-        disp_median_bin = disp_grouped.median()
-        # the next line raises the warning: "Mean of empty slice"
-        with warnings.catch_warnings():
-            warnings.simplefilter("ignore")
-            disp_mad_bin = disp_grouped.apply(robust.mad)
-            df["dispersions_norm"] = (
-                df["dispersions"].to_numpy()
-                - disp_median_bin[df["mean_bin"].to_numpy()].to_numpy()
-            ) / disp_mad_bin[df["mean_bin"].to_numpy()].to_numpy()
+    # actually do the normalization
+    df["dispersions_norm"] = (df["dispersions"] - disp_stats["avg"]) / disp_stats["dev"]
+    df["highly_variable"] = _subset_genes(
+        adata,
+        mean=mean,
+        dispersion_norm=df["dispersions_norm"].to_numpy(),
+        cutoff=cutoff,
+    )
+
+    df.set_index(adata.var_names, inplace=True)
+    df.index.name = "gene"
```

Comment:
I don't think we should be setting this

```suggestion
```

Reply:
sure, we can do that in `drop_index` instead